python/ray/serve/tests/unit/test_singleton_thread_router.py

```diff
@@ -0,0 +1,261 @@
+import asyncio
+import concurrent.futures
+import time
+from unittest.mock import Mock, patch
+
+import pytest
+
+from ray._private.test_utils import wait_for_condition
+from ray.serve._private.router import SingletonThreadRouter
+from ray.serve._private.common import DeploymentID, RequestMetadata
+from ray.serve._private.replica_result import ReplicaResult
+from ray.serve._private.router import AsyncioRouter
+
+
+def test_request_assignment():
+    mock_replica_result = Mock(spec=ReplicaResult)
+
+    mock_router = Mock(spec=AsyncioRouter)
+
+    async def mock_assign_request(*args, **kwargs):
+        return mock_replica_result
+
+    mock_router.assign_request.side_effect = mock_assign_request
+
+    with patch("ray.serve._private.router.AsyncioRouter", return_value=mock_router):
+        router = SingletonThreadRouter(
+            controller_handle=Mock(),
+            deployment_id=DeploymentID(name="test", app_name="test"),
+            handle_id="test",
+            self_actor_id="test",
+            handle_source="test",
+            replica_scheduler=Mock(),
+            enable_strict_max_ongoing_requests=False,
+            resolve_request_arg_func=Mock(),
+        )
+
+        request_meta = RequestMetadata(
+            request_id="test",
+            internal_request_id="test",
+        )
+
+        future = router.assign_request(request_meta)
+        assert isinstance(future, concurrent.futures.Future)
+        assert future.result() == mock_replica_result
+
+
+@pytest.mark.asyncio
+async def test_cancellation_propagation():
+    mock_replica_result = Mock(spec=ReplicaResult)
+    mock_replica_result.cancel = Mock()
+
+    loop = SingletonThreadRouter._get_singleton_asyncio_loop()
+
+    async def init_events():
+        return asyncio.Event(), asyncio.Event()
+
+    f = asyncio.run_coroutine_threadsafe(init_events(), loop)
+    lock_acquire_event, cancel_block_event = f.result()
```

Comment:
@alexeykudinkin let me know if there is a better way to do this. current implementation looks a bit messy because the event need to be on routers event loop.